iceoryx_examples/automotive_soa/include/owl/types.hpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_TYPES_HPP
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_TYPES_HPP
+
+#include "iceoryx_hoofs/cxx/function.hpp"
+#include "iceoryx_hoofs/cxx/optional.hpp"
+#include "iceoryx_hoofs/cxx/string.hpp"
+#include "iceoryx_posh/capro/service_description.hpp"
+#include "iceoryx_posh/popo/sample.hpp"
+
+#include "kom/sample_allocatee_ptr.hpp"
+
+#include <future>
+#include <memory>
+
+namespace owl
+{
+class Runtime;
+namespace core
+{
+template <typename T>
+using Result = T;
+
+using String = iox::cxx::string<100>;
+} // namespace core
+
+namespace kom
+{
+template <typename T>
+using SamplePtr = iox::cxx::optional<iox::popo::Sample<const T>>;
+
+using InstanceIdentifier = core::String;
+using ServiceIdentifier = core::String;
+
+struct FindServiceHandle
+{
+    // Only the Runtime shall be able to create handles, not the user
+    friend class owl::Runtime;
+
+  public:
+    const ServiceIdentifier& GetServiceId() const noexcept
+    {
+        return m_serviceIdentifier;
+    }
+    const InstanceIdentifier& GetInstanceId() const noexcept
+    {
+        return m_instanceIdentifier;
+    }
+
+  private:
+    FindServiceHandle(ServiceIdentifier serviceIdentifier, InstanceIdentifier instanceIdentifier) noexcept
+        : m_serviceIdentifier(serviceIdentifier)
+        , m_instanceIdentifier(instanceIdentifier)
+    {
+    }
+    ServiceIdentifier m_serviceIdentifier;
+    InstanceIdentifier m_instanceIdentifier;
```

Comment:
> Is this sufficient to uniquely identify the registered handler?

From an AUTOSAR point of view: Yes, there are only instances of services which can be uniquely identified.

> What if we register two handlers for the same instance?

Currently, allowed and will result in the callback being executed two times.

> Can we get an invalid identifier indicating that registration (`StartFindService`) failed?

AFAIK from the standard point of you, there is no possibility to communicate failure via a return value. I added a error print in case the vector is full.